File: packages/loader-api/src/api-loader.ts
import cache from "./cache";

const AGENTMARK_TEMPLATE_ENDPOINT = `v1/templates`;

/**
 * Prompt kind types supported by the API loader.
 */
export type PromptKind = "object" | "text" | "image" | "speech";

/**
 * Options for cloud mode - requires authentication credentials.
 */
export interface CloudLoaderOptions {
  apiKey: string;
  appId: string;
  baseUrl?: string;
}

/**
 * Options for local development mode - just needs a base URL.
 */
export interface LocalLoaderOptions {
  baseUrl: string;
}

/**
 * @deprecated Use `ApiLoader.cloud()` or `ApiLoader.local()` factory methods instead.
 * Legacy options interface for backward compatibility.
 */
export interface ApiLoaderOptions {
  apiKey: string;
  baseUrl: string;
  appId: string;
}

type FetchTemplateOptions = {
  cache:
    | {
        ttl?: number;
      }
    | false;
};

/**
 * ApiLoader fetches prompts from the AgentMark API or local dev server.
 *
 * @example Cloud mode (production)
 * ```ts
 * const loader = ApiLoader.cloud({
 *   apiKey: process.env.AGENTMARK_API_KEY!,
 *   appId: process.env.AGENTMARK_APP_ID!,
 * });
 * ```
 *
 * @example Local mode (development)
 * ```ts
 * const loader = ApiLoader.local({
 *   baseUrl: 'http://localhost:9418',
 * });
 * ```
 */
export class ApiLoader {
  private apiKey: string | undefined;
  private appId: string | undefined;
  private baseUrl: string;

  /**
   * @deprecated Use `ApiLoader.cloud()` or `ApiLoader.local()` factory methods instead.
   */
  constructor(options: ApiLoaderOptions) {
    this.apiKey = options.apiKey || undefined;
    this.appId = options.appId || undefined;
    this.baseUrl = options.baseUrl;
  }

  /**
   * Create a loader for cloud/production use with AgentMark API.
   * Requires API credentials.
   */
  static cloud(options: CloudLoaderOptions): ApiLoader {
    const instance = Object.create(ApiLoader.prototype);
    instance.apiKey = options.apiKey;
    instance.appId = options.appId;
    instance.baseUrl = options.baseUrl || 'https://api.agentmark.co';
    return instance;
  }

  /**
   * Create a loader for local development with the AgentMark dev server.
   * No credentials required - just point to your local server.
   */
  static local(options: LocalLoaderOptions): ApiLoader {
    const instance = Object.create(ApiLoader.prototype);
    instance.apiKey = undefined;
    instance.appId = undefined;
    instance.baseUrl = options.baseUrl;
    return instance;
  }

  async load(
    path: string,
    promptKind: PromptKind,
    options: FetchTemplateOptions = { cache: { ttl: 1000 * 60 } }
  ): Promise<unknown> {
    const ast =
      options.cache && cache.has(path as string)
        ? cache.get<any>(path as string)!
        : await this.fetchRequest({ path, promptKind });

    if (options.cache) {
      cache.set(path as string, ast, {
        ttl: options.cache.ttl,
      });
    }

    return ast;
  }

  async loadDataset(
    datasetPath: string
  ): Promise<
    ReadableStream<{ input: Record<string, unknown>; expected_output?: string }>
  > {
    const response = await this.fetchRequest(
      { path: datasetPath },
      { stream: true }
    );

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error("Failed to get response reader");
    }

    const decoder = new TextDecoder();
    let buffer = "";
    let rowCount = 0;

    return new ReadableStream({
      async pull(controller) {
        const { done, value } = await reader.read();

        if (done) {
          if (buffer.trim()) {
            try {
              const parsed = JSON.parse(buffer);
              // Handle case where server returns JSON array instead of NDJSON
              const items = Array.isArray(parsed) ? parsed : [parsed];
              for (const item of items) {
                // Validate that the row has an input field
                if (!item.input || typeof item.input !== 'object') {
                  throw new Error(`Invalid dataset row: missing or invalid 'input' field. Each row must have an 'input' object.`);
                }
                controller.enqueue(item);
                rowCount++;
              }
            } catch (e) {
              if (e instanceof SyntaxError) {
                throw new Error(`Failed to parse JSON in dataset ${datasetPath}: ${e.message}`);
              }
              throw e;
            }
          }
          if (rowCount === 0) {
            throw new Error(`Dataset ${datasetPath} is empty or contains no valid rows`);
          }
          controller.close();
          return;
        }

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split("\n");
        buffer = lines.pop() || "";

        for (const line of lines) {
          if (line.trim()) {
            try {
              const parsed = JSON.parse(line);
              // Handle case where server returns JSON array instead of NDJSON
              const items = Array.isArray(parsed) ? parsed : [parsed];
              for (const item of items) {
                // Validate that the row has an input field
                if (!item.input || typeof item.input !== 'object') {
                  throw new Error(`Invalid dataset row: missing or invalid 'input' field. Each row must have an 'input' object.`);
                }
                controller.enqueue(item);
                rowCount++;
              }
            } catch (e) {
              if (e instanceof SyntaxError) {
                throw new Error(`Failed to parse JSON in dataset ${datasetPath}: ${e.message}`);
              }
              throw e;
            }
          }
        }
      },
      cancel() {
        reader.cancel();
      },
    });
  }

  private async fetchRequest(
    queryParams: Record<string, string>,
    options: { stream?: boolean } = {}
  ) {
    const headers = new Headers({
      "Content-Type": "application/json",
    });

    // Only add auth headers in cloud mode (when credentials are present)
    if (this.apiKey && this.appId) {
      headers.set("X-Agentmark-App-Id", this.appId);
      headers.set("Authorization", this.apiKey);
    }

    // Request NDJSON format for streaming datasets
    if (options.stream) {
      headers.set("Accept", "application/x-ndjson");
    }

    const response = await fetch(
      `${this.baseUrl}/${AGENTMARK_TEMPLATE_ENDPOINT}?${new URLSearchParams(queryParams).toString()}`,
      {
        headers,
      }
    );

    if (response.ok) {
      if (options.stream) {
        return response;
      }

      return (await response.json()).data;
    }
    const errorResponse = await response.json();
    throw errorResponse.error;
  }
}

File: packages/prompt-core-python/src/agentmark/prompt_core/api_loader.py
"""API Loader for AgentMark prompt core.

Fetches prompts and datasets from the AgentMark API (cloud mode)
or a local dev server. Mirrors the TypeScript ApiLoader from @agentmark-ai/loader-api.
"""

from __future__ import annotations

import json
import os
from typing import Any

import httpx

from .types import DatasetStream, PromptKind


AGENTMARK_TEMPLATE_ENDPOINT = "v1/templates"


class ApiDatasetReader:
    """A dataset reader that reads items from a pre-fetched list."""

    def __init__(self, items: list[dict[str, Any]]) -> None:
        self._items = items
        self._index = 0

    async def read(self) -> dict[str, Any]:
        if self._index >= len(self._items):
            return {"done": True}
        item = self._items[self._index]
        self._index += 1
        return {"done": False, "value": item}


class ApiDatasetStream:
    """A dataset stream backed by an API response."""

    def __init__(self, items: list[dict[str, Any]]) -> None:
        self._items = items

    def get_reader(self) -> ApiDatasetReader:
        return ApiDatasetReader(self._items)


class ApiLoader:
    """Loader that fetches prompts and datasets from the AgentMark API.

    Matches the TypeScript ApiLoader from @agentmark-ai/loader-api.

    Example (cloud mode):
        loader = ApiLoader.cloud(
            api_key=os.environ["AGENTMARK_API_KEY"],
            app_id=os.environ["AGENTMARK_APP_ID"],
        )

    Example (local dev):
        loader = ApiLoader.local(base_url="http://localhost:9418")
    """

    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        app_id: str | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._app_id = app_id

    @classmethod
    def cloud(
        cls,
        api_key: str | None = None,
        app_id: str | None = None,
        base_url: str | None = None,
    ) -> ApiLoader:
        """Create a loader for cloud/production use.

        Falls back to AGENTMARK_API_KEY, AGENTMARK_APP_ID, and
        AGENTMARK_BASE_URL environment variables if not provided.
        """
        return cls(
            base_url=base_url or os.environ.get("AGENTMARK_BASE_URL", "https://api.agentmark.co"),
            api_key=api_key or os.environ.get("AGENTMARK_API_KEY"),
            app_id=app_id or os.environ.get("AGENTMARK_APP_ID"),
        )

    @classmethod
    def local(cls, base_url: str = "http://localhost:9418") -> ApiLoader:
        """Create a loader for local development."""
        return cls(base_url=base_url)

    async def load(
        self,
        path: str,
        prompt_type: PromptKind,
        options: dict[str, Any] | None = None,
    ) -> Any:
        """Load a prompt AST from the API."""
        data = await self._fetch_request(
            {"path": path, "promptKind": prompt_type}
        )
        return data

    async def load_dataset(self, dataset_path: str) -> DatasetStream:
        """Load a dataset from the API as a stream of JSONL rows."""
        response = await self._fetch_request(
            {"path": dataset_path},
            stream=True,
        )

        # Parse NDJSON lines into items
        items: list[dict[str, Any]] = []
        if isinstance(response, str):
            for line in response.strip().split("\n"):
                line = line.strip()
                if line:
                    parsed = json.loads(line)
                    if isinstance(parsed, list):
                        items.extend(parsed)
                    else:
                        items.append(parsed)
        elif isinstance(response, list):
            items = response

        if not items:
            raise ValueError(
                f"Dataset {dataset_path} is empty or contains no valid rows"
            )

        # Validate rows
        for item in items:
            if not isinstance(item.get("input"), dict):
                raise ValueError(
                    f"Invalid dataset row: missing or invalid 'input' field. "
                    f"Each row must have an 'input' object."
                )

        return ApiDatasetStream(items)

    async def _fetch_request(
        self,
        query_params: dict[str, str],
        stream: bool = False,
    ) -> Any:
        headers: dict[str, str] = {"Content-Type": "application/json"}

        # Auth headers for cloud mode
        if self._api_key and self._app_id:
            headers["X-Agentmark-App-Id"] = self._app_id
            headers["Authorization"] = self._api_key

        if stream:
            headers["Accept"] = "application/x-ndjson"

        url = f"{self._base_url}/{AGENTMARK_TEMPLATE_ENDPOINT}"

        async with httpx.AsyncClient() as client:
            response = await client.get(
                url,
                params=query_params,
                headers=headers,
                timeout=30.0,
            )

        if response.status_code == 200:
            if stream:
                return response.text
            return response.json().get("data")

        try:
            error_data = response.json()
            raise ValueError(error_data.get("error", f"API error: {response.status_code}"))
        except (json.JSONDecodeError, ValueError):
            raise ValueError(f"API error: {response.status_code} {response.text[:200]}")


__all__ = [
    "ApiLoader",
    "ApiDatasetStream",
    "ApiDatasetReader",
]
